File: BaseAndroidProject/app/src/main/java/com/lkb/baseandroidproject/PlayerActivity.java
package com.lkb.baseandroidproject;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.WindowManager;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;
public class PlayerActivity extends AppCompatActivity {
    private static final String ARG_VIDEO = "video";

    public static Intent createIntent(Context context, String videoUrl) {
        Intent intent = new Intent(context, PlayerActivity.class);
        intent.putExtra(ARG_VIDEO, videoUrl);
        return intent;
    }

    PlayerView mPlayerView;

    private SimpleExoPlayer mPlayer;
    private String mVideoUrl;

    private long mCurrentMillis;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mPlayerView = findViewById(R.id.full_screen_video);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
        loadArgs();
    }

    @Override
    protected void onResume() {
        super.onResume();
        startPlayer();
    }

    @Override
    protected void onPause() {
        release();
        super.onPause();
    }

    private void loadArgs() {
        if (getIntent() == null) {
            finish();
        }
        mVideoUrl = "assets:///heart_attack.mp3";
    }

    private void startPlayer() {
        if (mPlayer != null) {
            return;
        }
        mPlayer = ExoPlayerFactory.newSimpleInstance(new DefaultRenderersFactory(this),
                new DefaultTrackSelector());
        mPlayerView.setPlayer(mPlayer);

        DefaultDataSourceFactory dataSourceFactory =
                new DefaultDataSourceFactory(this, Util.getUserAgent(this, "player"));
        ExtractorMediaSource extractorMediaSource = new ExtractorMediaSource.Factory(dataSourceFactory).createMediaSource(Uri.parse(mVideoUrl));

        boolean isResuming = mCurrentMillis != 0;
        mPlayer.prepare(extractorMediaSource, isResuming, false);
        mPlayer.setPlayWhenReady(true);
        if (isResuming) {
            mPlayer.seekTo(mCurrentMillis);
        }

    }

    private void release() {
        if (mPlayer == null) {
            return;
        }
        mCurrentMillis = mPlayer.getCurrentPosition();
        mPlayer.release();
        mPlayer = null;
    }
}

File: BaseAndroidProject/app/src/main/java/com/lkb/baseandroidproject/PlayerActivity.kt
package com.lkb.baseandroidproject

import android.annotation.SuppressLint
import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.google.android.exoplayer2.DefaultLoadControl
import com.google.android.exoplayer2.DefaultRenderersFactory
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.ExoPlayerFactory
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory
import com.google.android.exoplayer2.util.Util
import kotlinx.android.synthetic.main.activity_main2.*

class PlayerActivity:AppCompatActivity(){
    var exoPlayer: ExoPlayer? = null
    var mPlayerCurrentPosition: Long? = 0
    var mPlayerCurrentWindow: Int? = 0
    var mPlayerReady: Boolean? = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main2)
    }

    /*
          Combinning ExoPlayer with Activity LifeCycle
       */
    public override fun onStart() {
        super.onStart()
        if (Util.SDK_INT > 23) {
            initializePlayer()
        }
    }

    public override fun onResume() {
        super.onResume()
        hideSystemUi()
        if (Util.SDK_INT <= 23 || exoPlayer == null) {
            initializePlayer()
        }
    }

    public override fun onPause() {
        super.onPause()
        if (Util.SDK_INT <= 23) {
            releasePlayer()
        }
    }

    public override fun onStop() {
        super.onStop()
        if (Util.SDK_INT > 23) {
            releasePlayer()
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        if (exoPlayer != null)
            exoPlayer?.release()
        exoPlayer = null
    }

    // This function is responsible to initialize ExoPlayer.
    private fun initializePlayer() {
        if (exoPlayer == null) {
            exoPlayer = ExoPlayerFactory.newSimpleInstance(
                DefaultRenderersFactory(this),
                DefaultTrackSelector(), DefaultLoadControl()
            )
            video_view?.player = exoPlayer
            video_view?.player?.playWhenReady = true
        }

        val uri = Uri.parse(getString(R.string.exoplayer_media_source))
        exoPlayer?.prepare(buildMediaSource(uri), false, false)
        video_view?.player?.seekTo(mPlayerCurrentWindow!!, mPlayerCurrentPosition!!)
    }

    // This function is responsible to relase framework resources when not needed.
    private fun releasePlayer() {
        if (exoPlayer != null) {
            mPlayerCurrentPosition = exoPlayer?.currentPosition
            mPlayerCurrentWindow = exoPlayer?.currentWindowIndex
            mPlayerReady = exoPlayer?.playWhenReady
            exoPlayer?.release()
        }
    }
    /*
  * HlsMediaSource Builduing from given URI (video source).
  * */
    private fun buildMediaSource(uri: Uri): MediaSource {
        return HlsMediaSource.Factory(
            DefaultHttpDataSourceFactory("exoplayer")
        )
            .setAllowChunklessPreparation(true)
            .createMediaSource(uri)
    }

    // To have full screen streeming expereince in immersive mode.
    @SuppressLint("InlinedApi")
    private fun hideSystemUi() {
        video_view.setSystemUiVisibility(
            View.SYSTEM_UI_FLAG_LOW_PROFILE
                    or View.SYSTEM_UI_FLAG_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
    }
}